contracts/examples/erc1155/src/lib.rs

```diff
@@ -0,0 +1,527 @@
+		for (i, type_id) in type_ids.iter().enumerate() {
+			if self.get_is_fungible(type_id) == true {
+				let amount = &values[i];
+
+				let balance = match self.get_balance_mapper(&from).get(type_id) {
+					Some(b) => b,
+					None => return sc_error!("Address has no tokens of that type"),
+				};
+
+				require!(amount > &0, "Must transfer more than 0");
+				require!(amount <= &balance, "Not enough balance for id");
+
+				self.decrease_balance(&from, &type_id, &amount);
+
+				if !is_receiver_smart_contract {
+					self.increase_balance(&to, &type_id, &amount);
+				}
+			} else {
+				let token_id = &values[i];
+
+				require!(
+					self.get_token_owner(&type_id, &token_id) == from,
+					"_from_ is not the owner of the token"
+				);
+
+				let amount = BigUint::from(1u32);
+				self.decrease_balance(&from, &type_id, &amount);
+
+				if !is_receiver_smart_contract {
+					self.increase_balance(&to, &type_id, &amount);
+					self.set_token_owner(&type_id, &token_id, &to);
+				} else {
+					self.set_token_owner(&type_id, &token_id, &Address::zero());
+				}
+			}
+		}
+
+		if is_receiver_smart_contract {
+			self.peform_async_call_batch_transfer(from, to, type_ids, values, data);
+		}
+
+		// self.transfer_batch_event(&caller, &from, &to, ids, amounts);
+
+		Ok(())
+	}
+
+	#[endpoint(setApprovalForAll)]
+	fn set_approved_for_all(&self, operator: Address, approved: bool) {
+		let caller = self.get_caller();
+
+		self.set_is_approved(&operator, &caller, approved);
+
+		// self.approval_for_all_event(&caller, &operator, approved);
+	}
+
+	// returns assigned id
+	#[endpoint(createToken)]
+	fn create_token(&self, uri: &[u8], initial_supply: BigUint, is_fungible: bool) -> BigUint {
+		let big_uint_one = BigUint::from(1u32);
+
+		let creator = self.get_caller();
+		let type_id = self.get_last_valid_type_id() + big_uint_one.clone();
+
+		self.get_balance_mapper(&creator)
+			.insert(type_id.clone(), initial_supply.clone());
+		self.set_token_type_creator(&type_id, &creator);
+		self.set_is_fungible(&type_id, is_fungible);
+
+		if !is_fungible {
+			self.set_owner_for_range(&type_id, &big_uint_one, &initial_supply, &creator);
+			self.set_last_valid_token_id_for_type(&type_id, &initial_supply);
+		}
+
+		self.set_last_valid_type_id(&type_id);
+		self.set_token_type_uri(&type_id, uri);
+
+		// self.transfer_single_event(&caller, &from, &to, &id, &amount);
+
+		// uri event
+
+		type_id
+	}
+
+	#[endpoint]
+	fn mint(&self, type_id: BigUint, amount: BigUint) -> SCResult<()> {
+		let creator = self.get_token_type_creator(&type_id);
+
+		require!(
+			self.get_caller() == creator,
+			"Only the token creator may mint more tokens"
+		);
+
+		self.increase_balance(&creator, &type_id, &amount);
+
+		if self.get_is_fungible(&type_id) == false {
+			let id_first = self.get_last_valid_token_id_for_type(&type_id);
+			let id_last = id_first.clone() + amount;
+
+			self.set_owner_for_range(&type_id, &id_first, &id_last, &creator);
```

Comment:
Bug: self.set_owner_for_range sets the owner for the range including both start and end, but here the tokens generated are only from start + 1 to end.
For example in the mandos test, if tokens 1-5 already exist, and 3 tokens are added, only the owner for tokens 6, 7, 8 should be set.
Fix: add 1 to let id_first, and subtract one from id_last
Additionally, change the mandos test so that it checks that the owner of the first 5 tokens is untouched (eg. first transfer them to someone else).